File: backend/models/response/scrape.go
package scrape

import (
	. "github.com/martynasd123/golang-scraper/storage"
)

type AddTaskResponse struct {
	Id int `json:"id"`
}

func CreateAddTaskResponse(id int) *AddTaskResponse {
	return &AddTaskResponse{id}
}

type TaskStatusResponse struct {
	Id                *int    `json:"id"`
	Status            string  `json:"status"`
	Link              string  `json:"link"`
	ExternalLinks     *int    `json:"externalLinks"`
	InternalLinks     *int    `json:"internalLinks"`
	InaccessibleLinks *int    `json:"inaccessibleLinks"`
	HtmlVersion       *string `json:"htmlVersion"`
	PageTitle         *string `json:"pageTitle"`
	HeadingsByLevel   *[6]int `json:"headingsByLevel"`
	CrawledLinks      int     `json:"crawledLinks"`
	Error             *string `json:"error"`
}

func CreateTaskStatusResponse(task *Task) *TaskStatusResponse {
	response := &TaskStatusResponse{}
	response.Id = task.Id
	response.Status = task.Status
	response.InaccessibleLinks = task.InaccessibleLinks
	response.HtmlVersion = task.HtmlVersion
	response.PageTitle = task.PageTitle
	response.HeadingsByLevel = task.HeadingsByLevel
	response.Link = task.Link.String()
	response.ExternalLinks = task.ExternalLinks
	response.InternalLinks = task.InternalLinks
	response.CrawledLinks = task.CrawledLinks
	response.Error = task.Error
	return response
}

File: frontend/src/api/scrape.ts
import {AuthenticatedClient, Client} from "./client";
import {AxiosResponse} from "axios";

interface AddTaskRequest {
    link: string
}

interface AddTaskResponse {
    id: number
}

interface TaskListItemResponse {
    id: number;
    link: string;
    inaccessibleLinks?: number;
    pageTitle?: string;
    crawledLinks?: number;
    error?: string;
    status: string;
}

export enum TaskStatus {
    STATUS_PENDING = "PENDING",
    STATUS_INITIATING = "INITIATING",
    STATUS_TRYING_LINKS = "TRYING_LINKS",
    STATUS_FINISHED = "FINISHED",
    STATUS_ERROR = "ERROR",
    STATUS_INTERRUPTED = "INTERRUPTED",
    STATUS_INTERRUPTING = "INTERRUPTING"
}

export interface TaskStateUpdate {
    id?: number;
    status: TaskStatus;
    link: string;
    externalLinks?: number;
    internalLinks?: number;
    inaccessibleLinks?: number;
    htmlVersion?: string;
    pageTitle?: string;
    headingsByLevel?: [number, number, number, number, number, number];
    crawledLinks: number;
    error?: string;
}

export const sendAddTaskRequest = (request: AddTaskRequest) => AuthenticatedClient
    .post<AddTaskRequest, AxiosResponse<AddTaskResponse>>("/api/scrape/add-task", request)

export const sendInterruptTaskRequest = (id: string) => AuthenticatedClient
    .post<void, AxiosResponse<void>>(`/api/scrape/task/${id}/interrupt`)

export const sendGetTasksRequest = () => AuthenticatedClient
    .get<AddTaskRequest, AxiosResponse<TaskListItemResponse[]>>("/api/scrape/tasks")
